File: src/webapp/src/pulse/patterns.ts
// Animation patterns. Each pattern has:
//   init(size, rng) -> params  (called once, may randomize via the supplied rng)
//   sample(t, x, y, params, time) -> intensity in [0, 1]
// `t` is loop-normalized in [0, 1) and loops seamlessly at the boundary.
// `time` is monotonic real seconds; only patterns with per-pixel state need it.
// All randomness flows from the `Rand` passed into init (and stored on params
// when a pattern needs more randomness during sampling), so a given seed
// reproduces the same visual.

import type { Rand } from './rng';

const TWO_PI = Math.PI * 2;

const range = (rng: Rand, min: number, max: number) => min + rng() * (max - min);
const pick = <T>(rng: Rand, arr: readonly T[]): T => arr[Math.floor(rng() * arr.length)];

// Value noise --------------------------------------------------------------

function hash2(x: number, y: number, seed: number): number {
  let h = Math.imul(x | 0, 374761393) + Math.imul(y | 0, 668265263) + Math.imul(seed | 0, 982451653);
  h = Math.imul(h ^ (h >>> 13), 1274126177);
  return ((h ^ (h >>> 16)) >>> 0) / 4294967295;
}
const smooth = (t: number) => t * t * (3 - 2 * t);
function valueNoise(x: number, y: number, seed: number): number {
  const xi = Math.floor(x), yi = Math.floor(y);
  const xf = x - xi, yf = y - yi;
  const a = hash2(xi, yi, seed);
  const b = hash2(xi + 1, yi, seed);
  const c = hash2(xi, yi + 1, seed);
  const d = hash2(xi + 1, yi + 1, seed);
  const u = smooth(xf), v = smooth(yf);
  return a * (1 - u) * (1 - v) + b * u * (1 - v) + c * (1 - u) * v + d * u * v;
}
function fbm(x: number, y: number, seed: number, octaves: number): number {
  let sum = 0, amp = 1, total = 0, freq = 1;
  for (let i = 0; i < octaves; i++) {
    sum += valueNoise(x * freq, y * freq, seed + i * 17) * amp;
    total += amp;
    amp *= 0.5;
    freq *= 2;
  }
  return sum / total;
}

// Pattern interface --------------------------------------------------------

export interface Pattern<P = unknown> {
  init(size: number, rng: Rand): P;
  sample(t: number, x: number, y: number, params: P, time: number): number;
}

// Patterns -----------------------------------------------------------------

interface PulseParams { cx: number; cy: number; waveWidth: number; ripples: number; maxRadius: number; }
const pulse: Pattern<PulseParams> = {
  init(size, rng) {
    const waveWidth = range(rng, 1.8, 3.2);
    return {
      cx: range(rng, size / 2 - 1, size / 2 + 1),
      cy: range(rng, size / 2 - 1, size / 2 + 1),
      waveWidth,
      ripples: pick(rng, [1, 2]),
      maxRadius: Math.hypot(size, size) / 2 + waveWidth,
    };
  },
  sample(t, x, y, p, _time) {
    const dist = Math.hypot(x - p.cx + 0.5, y - p.cy + 0.5);
    let v = 0;
    for (let r = 0; r < p.ripples; r++) {
      const rPhase = (t + r / p.ripples) % 1;
      const waveR = rPhase * p.maxRadius;
      const delta = Math.abs(dist - waveR);
      const falloff = Math.max(0, 1 - delta / p.waveWidth);
      const trail = dist < waveR ? 0.7 : 1.0;
      v = Math.max(v, falloff * trail);
    }
    return v;
  },
};

interface NoiseParams { scale: number; loopR: number; seed: number; octaves: number; }
const noise: Pattern<NoiseParams> = {
  init(_size, rng) {
    return {
      scale: range(rng, 0.18, 0.32),
      loopR: range(rng, 0.6, 1.4),
      seed: Math.floor(rng() * 1e6),
      octaves: 3,
    };
  },
  sample(t, x, y, p, _time) {
    const ox = Math.cos(TWO_PI * t) * p.loopR;
    const oy = Math.sin(TWO_PI * t) * p.loopR;
    const n = fbm(x * p.scale + ox, y * p.scale + oy, p.seed, p.octaves);
    return Math.max(0, Math.min(1, (n - 0.2) / 0.6));
  },
};

interface WavesParams { angle: number; freq: number; speed: number; sharpness: number; cx: number; cy: number; }
const waves: Pattern<WavesParams> = {
  init(size, rng) {
    return {
      angle: range(rng, 0, TWO_PI),
      freq: range(rng, 0.6, 1.4),
      speed: pick(rng, [1, -1]),
      sharpness: range(rng, 1.2, 2.4),
      cx: size / 2 - 0.5,
      cy: size / 2 - 0.5,
    };
  },
  sample(t, x, y, p, _time) {
    const proj = (x - p.cx) * Math.cos(p.angle) + (y - p.cy) * Math.sin(p.angle);
    const s = Math.sin(proj * p.freq + p.speed * TWO_PI * t);
    return Math.pow((s + 1) / 2, p.sharpness);
  },
};

interface FlowParams { angle: number; bandWidth: number; bands: number; period: number; speed: number; cx: number; cy: number; }
const flow: Pattern<FlowParams> = {
  init(size, rng) {
    return {
      angle: range(rng, 0, TWO_PI),
      bandWidth: range(rng, 2.5, 4.5),
      bands: pick(rng, [2, 3]),
      period: range(rng, 10, 16),
      speed: pick(rng, [1, -1]),
      cx: size / 2 - 0.5,
      cy: size / 2 - 0.5,
    };
  },
  sample(t, x, y, p, _time) {
    const proj = (x - p.cx) * Math.cos(p.angle) + (y - p.cy) * Math.sin(p.angle);
    const norm = ((proj + p.speed * t * p.period) % p.period + p.period) % p.period;
    let v = 0;
    for (let b = 0; b < p.bands; b++) {
      const bandPos = (b * p.period) / p.bands;
      const raw = Math.abs(norm - bandPos);
      const d = Math.min(raw, p.period - raw);
      v = Math.max(v, Math.max(0, 1 - d / p.bandWidth));
    }
    return v;
  },
};

interface SpiralParams { arms: number; tightness: number; speed: number; sharpness: number; cx: number; cy: number; }
const spiral: Pattern<SpiralParams> = {
  init(size, rng) {
    return {
      arms: pick(rng, [1, 2, 3]),
      tightness: range(rng, 0.5, 1.1),
      speed: pick(rng, [1, -1]),
      sharpness: range(rng, 1.4, 2.4),
      cx: size / 2 - 0.5,
      cy: size / 2 - 0.5,
    };
  },
  sample(t, x, y, p, _time) {
    const dx = x - p.cx, dy = y - p.cy;
    const theta = Math.atan2(dy, dx);
    const dist = Math.hypot(dx, dy);
    const s = Math.sin(theta * p.arms + dist * p.tightness - p.speed * TWO_PI * t);
    return Math.pow((s + 1) / 2, p.sharpness);
  },
};

interface PlasmaParams { a: number; b: number; c: number; d: number; sa: number; sb: number; sc: number; sd: number; cx: number; cy: number; }
const plasma: Pattern<PlasmaParams> = {
  init(size, rng) {
    return {
      a: range(rng, 0.35, 0.7),
      b: range(rng, 0.35, 0.7),
      c: range(rng, 0.25, 0.6),
      d: range(rng, 0.4, 0.8),
      sa: pick(rng, [1, -1, 2, -2]),
      sb: pick(rng, [1, -1, 2]),
      sc: pick(rng, [1, -1]),
      sd: pick(rng, [1, 2]),
      cx: size / 2 - 0.5,
      cy: size / 2 - 0.5,
    };
  },
  sample(t, x, y, p, _time) {
    const phase = TWO_PI * t;
    const v = (
      Math.sin(x * p.a + phase * p.sa) +
      Math.sin(y * p.b + phase * p.sb) +
      Math.sin((x + y) * p.c + phase * p.sc) +
      Math.sin(Math.hypot(x - p.cx, y - p.cy) * p.d + phase * p.sd)
    ) / 4;
    return (v + 1) / 2;
  },
};

interface ShimmerCell { from: number; to: number; startTime: number; duration: number; }
interface ShimmerParams { pixels: ShimmerCell[][]; minDur: number; maxDur: number; rng: Rand; }
const shimmer: Pattern<ShimmerParams> = {
  init(size, rng) {
    const minDur = 0.9;
    const maxDur = 2.8;
    const pixels: ShimmerCell[][] = [];
    for (let y = 0; y < size; y++) {
      const row: ShimmerCell[] = [];
      for (let x = 0; x < size; x++) {
        const duration = range(rng, minDur, maxDur);
        row.push({
          from: rng(),
          to: rng(),
          startTime: -range(rng, 0, duration),
          duration,
        });
      }
      pixels.push(row);
    }
    return { pixels, minDur, maxDur, rng };
  },
  sample(_t, x, y, p, time) {
    const px = p.pixels[y][x];
    while (time - px.startTime >= px.duration) {
      px.startTime += px.duration;
      px.from = px.to;
      px.to = p.rng();
      px.duration = range(p.rng, p.minDur, p.maxDur);
    }
    const progress = (time - px.startTime) / px.duration;
    const eased = progress * progress * (3 - 2 * progress);
    return px.from + (px.to - px.from) * eased;
  },
};

export const PATTERNS = { pulse, noise, waves, flow, spiral, plasma, shimmer } as const;
export type PatternName = keyof typeof PATTERNS;

export function pickPattern(rng: Rand, name?: PatternName): { name: PatternName; def: Pattern } {
  const keys = Object.keys(PATTERNS) as PatternName[];
  const chosen = name && PATTERNS[name] ? name : keys[Math.floor(rng() * keys.length)];
  return { name: chosen, def: PATTERNS[chosen] as Pattern };
}

File: src/api/Gabriel.Engine/Sequence/Patterns.cs
namespace Gabriel.Engine.Sequence;

// Pattern primitives ported (with tweaks) from the webapp's pulse/patterns.ts.
// Each pattern takes a coordinate (x, y) and a normalized time t in [0, 1) +
// pattern-specific parameters; returns a scalar in [0, 1] that the generator
// maps to a palette index.
//
// Patterns are designed for LOOP CONTINUITY — sampling at t=0 and t=1 should
// produce the same image so the 64-frame cycle closes seamlessly.
internal static class Patterns
{
    private const double TwoPi = Math.PI * 2;

    // --- Plasma ----------------------------------------------------------------
    //
    // Superposed sines in x, y, x+y, and radial-distance. Gorgeous out-of-the-box
    // animation, hugely varied across the parameter space — small seed changes
    // produce noticeably different visuals.

    public readonly record struct PlasmaParams(
        double A, double B, double C, double D,
        double Sa, double Sb, double Sc, double Sd,
        double Cx, double Cy);

    public static PlasmaParams PlasmaInit(Random rng, int size)
    {
        return new PlasmaParams(
            A: Range(rng, 0.35, 0.70),
            B: Range(rng, 0.35, 0.70),
            C: Range(rng, 0.25, 0.60),
            D: Range(rng, 0.40, 0.80),
            Sa: Pick(rng, new[] { 1.0, -1.0, 2.0, -2.0 }),
            Sb: Pick(rng, new[] { 1.0, -1.0, 2.0 }),
            Sc: Pick(rng, new[] { 1.0, -1.0 }),
            Sd: Pick(rng, new[] { 1.0, 2.0 }),
            Cx: size / 2.0 - 0.5,
            Cy: size / 2.0 - 0.5);
    }

    public static double Plasma(double x, double y, double t, in PlasmaParams p)
    {
        var phase = TwoPi * t;
        var v = (
            Math.Sin(x * p.A + phase * p.Sa) +
            Math.Sin(y * p.B + phase * p.Sb) +
            Math.Sin((x + y) * p.C + phase * p.Sc) +
            Math.Sin(Math.Sqrt((x - p.Cx) * (x - p.Cx) + (y - p.Cy) * (y - p.Cy)) * p.D + phase * p.Sd)
        ) / 4.0;
        return (v + 1) / 2;
    }

    // --- Directional waves -----------------------------------------------------

    public readonly record struct WavesParams(
        double Angle, double Freq, double Speed, double Sharpness, double Cx, double Cy);

    public static WavesParams WavesInit(Random rng, int size)
    {
        return new WavesParams(
            Angle: Range(rng, 0, TwoPi),
            Freq: Range(rng, 0.6, 1.4),
            Speed: rng.NextDouble() < 0.5 ? 1.0 : -1.0,
            Sharpness: Range(rng, 1.2, 2.4),
            Cx: size / 2.0 - 0.5,
            Cy: size / 2.0 - 0.5);
    }

    public static double Waves(double x, double y, double t, in WavesParams p)
    {
        var proj = (x - p.Cx) * Math.Cos(p.Angle) + (y - p.Cy) * Math.Sin(p.Angle);
        var s = Math.Sin(proj * p.Freq + p.Speed * TwoPi * t);
        return Math.Pow((s + 1) / 2, p.Sharpness);
    }

    // --- Spiral ----------------------------------------------------------------

    public readonly record struct SpiralParams(
        int Arms, double Tightness, double Speed, double Sharpness, double Cx, double Cy);

    public static SpiralParams SpiralInit(Random rng, int size)
    {
        return new SpiralParams(
            Arms: Pick(rng, new[] { 1, 2, 3 }),
            Tightness: Range(rng, 0.5, 1.1),
            Speed: rng.NextDouble() < 0.5 ? 1.0 : -1.0,
            Sharpness: Range(rng, 1.4, 2.4),
            Cx: size / 2.0 - 0.5,
            Cy: size / 2.0 - 0.5);
    }

    public static double Spiral(double x, double y, double t, in SpiralParams p)
    {
        var dx = x - p.Cx;
        var dy = y - p.Cy;
        var theta = Math.Atan2(dy, dx);
        var dist = Math.Sqrt(dx * dx + dy * dy);
        var s = Math.Sin(theta * p.Arms + dist * p.Tightness - p.Speed * TwoPi * t);
        return Math.Pow((s + 1) / 2, p.Sharpness);
    }

    // --- Pulse (expanding ripples) --------------------------------------------

    public readonly record struct PulseParams(
        double Cx, double Cy, double WaveWidth, int Ripples, double MaxRadius);

    public static PulseParams PulseInit(Random rng, int size)
    {
        var waveWidth = Range(rng, 1.8, 3.2);
        return new PulseParams(
            Cx: Range(rng, size / 2.0 - 1, size / 2.0 + 1),
            Cy: Range(rng, size / 2.0 - 1, size / 2.0 + 1),
            WaveWidth: waveWidth,
            Ripples: Pick(rng, new[] { 1, 2 }),
            MaxRadius: Math.Sqrt(size * size + size * size) / 2 + waveWidth);
    }

    public static double Pulse(double x, double y, double t, in PulseParams p)
    {
        var dist = Math.Sqrt((x - p.Cx + 0.5) * (x - p.Cx + 0.5) + (y - p.Cy + 0.5) * (y - p.Cy + 0.5));
        var v = 0.0;
        for (var r = 0; r < p.Ripples; r++)
        {
            var rPhase = (t + (double)r / p.Ripples) % 1.0;
            var waveR = rPhase * p.MaxRadius;
            var delta = Math.Abs(dist - waveR);
            var falloff = Math.Max(0, 1 - delta / p.WaveWidth);
            var trail = dist < waveR ? 0.7 : 1.0;
            v = Math.Max(v, falloff * trail);
        }
        return v;
    }

    // --- Shimmer (per-pixel independent phase) --------------------------------
    //
    // The key to the "glorious shimmer" feel: every pixel has its own random
    // phase offset, so they all oscillate independently. The animation looks
    // like a starfield rather than flowing tissue.

    public readonly record struct ShimmerParams(double[] PhaseOffsets, double[] Speeds, double Floor);

    public static ShimmerParams ShimmerInit(Random rng, int size)
    {
        var n = size * size;
        var phaseOffsets = new double[n];
        var speeds = new double[n];
        for (var i = 0; i < n; i++)
        {
            phaseOffsets[i] = rng.NextDouble();
            speeds[i] = 0.7 + rng.NextDouble() * 1.6;
        }
        return new ShimmerParams(phaseOffsets, speeds, Floor: 0.10);
    }

    public static double Shimmer(int x, int y, int size, double t, in ShimmerParams p)
    {
        var idx = y * size + x;
        var s = Math.Sin(TwoPi * (t * p.Speeds[idx] + p.PhaseOffsets[idx]));
        var v = (s + 1) / 2;
        return p.Floor + v * (1 - p.Floor);
    }

    // --- helpers ---------------------------------------------------------------

    private static double Range(Random rng, double min, double max) => min + rng.NextDouble() * (max - min);

    private static T Pick<T>(Random rng, T[] arr) => arr[rng.Next(arr.Length)];
}
